docs/engineering/configurations.md
### Configurations

Each of the configs should be namespaced according to the role/usage/context that they have followed by the type/implementation of it. For example - for statsd url - statsd is an implementation for metric collection hence the config would look like metric.statsd.url
Any delimiters depending on the language/framework can be used. Like “.” or “\_” or “/” etc. Default would be `.`

| Property Name                                   | Namespace       | Type (optional) | Config Name                        |
| ----------------------------------------------- | --------------- | --------------- | ---------------------------------- |
| metric.statsd.url                               | metric          | statsd          | url                                |
| schema.proto.name                               | schema          | proto           | name                               |
| dlq.kafka.brokers                               | dlq             | kafka           | brokers                            |
| dlq.enable                                      | dlq             | -               | enable                             |
| source.kafka.consumer.config.auto.commit.enable | source          | kafka           | consumer.config.auto.commit.enable |
| schema.registry.stencil.urls                    | schema.registry | stencil         | urls                               |

These are just guidelines and might not be extensive enough to cover all use-cases. Please use this to give it a thought.

### Metrics

The metrics and labels must be compliant with the [Prometheus data model](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels).
The metric and label conventions presented in this document serve as both a;[style-guide](https://prometheus.io/docs/practices/naming/) and a collection of best practices.


go-app/docs/reference/configuration.md
# Configuration

| Go struct                     | YAML path        | ENV var          | default   | Valid values                                                                                                     |
| ----------------------------- | ---------------- | ---------------- | --------- | ---------------------------------------------------------------------------------------------------------------- |
| Config.LogConfig.Level        | log.level        | LOG_LEVEL        | info      | debug,info,warn,error,dpanic,panic,fatal                                                                         |
| Config.Port                   | port             | PORT             | 8080      | 0-65535                                                                                                          |
| Config.DBConfig.User          | db.user          | DB_USER          | postgres  | [PostgreSQL identifiers](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS) |
| Config.DBConfig.Name          | db.name          | DB_NAME          | postgres  | [PostgreSQL identifiers](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS) |
| Config.DBConfig.Port          | db.port          | DB_PORT          | 5432      | 0-65535                                                                                                          |
| Config.DBConfig.Password      | db.password      | DB_PASSWORD      |           | valid PostgreSQL password                                                                                        |
| Config.DBConfig.SslMode       | db.sslmode       | DB_SSLMODE       | disable   | [libpq sslmode](https://www.postgresql.org/docs/9.1/libpq-ssl.html)                                              |
| Config.DBConfig.LogLevel      | db.log_level     | DB_LOG_LEVEL     | info      | silent,error,warn,info                                                                                           |
| Config.DBConfig.Host          | db.host          | DB_HOST          | localhost | valid hostname name or IP address                                                                                |
| Config.NewRelicConfig.Enabled | newrelic.enabled | NEWRELIC_ENABLED | false     | bool                                                                                                             |
| Config.NewRelicConfig.License | newrelic.license | NEWRELIC_LICENSE |           | 40 char NewRelic license key                                                                                     |
| Config.NewRelicConfig.AppName | newrelic.appname | NEWRELIC_APPNAME | app       | string                                                                                                           |

## How to configure

There are 3 ways to configure app:

- Using env variables
- Using a yaml file
- or using a combination of both

### Using env variables

Example:

```sh
export PORT=9999
go run main.go serve
```

This will run the service on port 9999 instead of the default 8080

### Using a yaml file

For default values and the structure of the yaml file please check file - [config.yaml.example](config.yaml.example)

Usage example:

```sh
cp config.yaml.example config.yaml
# make any modifications to the configs as required
go run main.go serve
```

### Using a combinnation of both

If any key that is set via both env vars and yaml the value set in env vars will take effect.


CHANGELOG.md
# CHANGELOG

These change logs could be a link to wherever you're maintaing the change logs. eg, Git tags or an automated way to update this file itself.


go-app/domain/config.go
package domain

// DBConfig contains the database configuration
type DBConfig struct {
	Host     string `mapstructure:"host" default:"localhost"`
	User     string `mapstructure:"user" default:"postgres"`
	Password string `mapstructure:"password" default:""`
	Name     string `mapstructure:"name" default:"postgres"`
	Port     string `mapstructure:"port" default:"5432"`
	SslMode  string `mapstructure:"sslmode" default:"disable"`
	LogLevel string `mapstructure:"log_level" default:"info"`
}

type AppServiceConfig struct {
	Host string `mapstructure:"host" default:"http://localhost:3000"`
}

// NewRelic contains the New Relic go-agent configuration
type NewRelicConfig struct {
	Enabled bool   `mapstructure:"enabled" default:"false"`
	AppName string `mapstructure:"appname" default:"app"`
	License string `mapstructure:"license"`
}

type LogConfig struct {
	Level string `mapstructure:"level" default:"info"`
}

// Config contains the application configuration
type Config struct {
	Port       int              `mapstructure:"port" default:"8080"`
	DB         DBConfig         `mapstructure:"db"`
	NewRelic   NewRelicConfig   `mapstructure:"newrelic"`
	AppService AppServiceConfig `mapstructure:"app_service"`
	Log        LogConfig        `mapstructure:"log"`
}
